### PR TITLE
Cancel ops in messenger when ops are redirected.

### PR DESCRIPTION
**Pull request by liyichao**
Separated from https://github.com/ceph/ceph/pull/24943 , some backgrounds are at https://github.com/ceph/ceph/pull/12216 .

* Cancel ops to the old osd when ops are redirected, so buffer will not be used after it has been freed.
* To cancel ops already in outcoming_bl, we deep copy data part of the message so it will not reference the buffer which may have been freed.
* out_seq is only updated for new messages, and messages from sent queue are sent using their original sequence. So even when we remove ops from sent queue, the out_seq is correct.

### CODE REVIEW

**Comment by liyichao** [comment]
Yeah, in simple messenger case, it will block, but it is acceptable because if that is a problem, the user can switch to asyncMessenger anyway, it will take many efforts to make it efficient so we just make it correct.

I will update to clone the whole `outcoming_bl`. 

As to the `write_lock`,  if we reuse `write_lock`, in normal operation, the `_try_send` in `write_message` will block `send_message`, reducing performance. Using another lock, the `_try_send` will only block `cancel_ops`.

File: src/msg/async/AsyncConnection.cc [hunk 1]
```diff
@@ -283,11 +283,12 @@ ssize_t AsyncConnection::read_bulk(char *buf, unsigned len)
 ssize_t AsyncConnection::write(bufferlist &bl,
                                std::function<void(ssize_t)> callback,
                                bool more) {
-
-    std::unique_lock<std::mutex> l(write_lock);
+    send_lock.lock();
```
Comment (liyichao) [inline src/msg/async/AsyncConnection.cc]: There is no need to protect _try_send and outcoming_bl with write_lock, see `ProtocolV1::write_event` for example.

**Comment by liyichao** [comment]
retest this please

**Comment by liyichao** [comment]
retest this please

**Comment by liyichao** [comment]
retest this please

**Comment by liyichao** [comment]
ping @dillaman

**Comment by liyichao** [comment]
retest this please

**Comment by liyichao** [comment]
retest this please

**Comment by dillaman** [comment]
@gregsfortytwo This allows librbd (and others) to avoid an extra memory copy when going from a C API. It actually makes a very noticeable performance impact for larger IO.

**Comment by liyichao** [comment]
> @gregsfortytwo This allows librbd (and others) to avoid an extra memory copy when going from a C API. It actually makes a very noticeable performance impact for larger IO.

Some concrete numbers: cpu flame graph shows 40% CPU is in the buffer::copy, and this will remove the copy.
image

**Comment by gregsfortytwo** [comment]
Can you talk through that a bit here then? It's not mentioned as a motivation in the commits or comments I see when I skim through for an overview, and what is present sets off a lot of alarms in my head for being very high-risk. :)

**Comment by liyichao** [comment]
> Can you talk through that a bit here then? It's not mentioned as a motivation in the commits or comments I see when I skim through for an overview, and what is present sets off a lot of alarms in my head for being very high-risk. :)

Sorry about that, This PR is separated from #24943 ,  some backgrounds are at #12216 .

**Comment by dillaman** [comment]
Note: PR #26932 re-implements the outbound message queue, which would conflict w/ this PR.

**Comment by rouming** [comment]
@liyichao 
> To cancel ops already in outcoming_bl, we deep copy data part of the message so it will not reference the buffer which may have been freed.

Seems to me this is wrong from the very beginning. All you need is to keep messages queued and put them only when data is sent. Then you do not need to copy anything. Just take the messages which are queued but not yet sent. This is O(1) operation, since all queued messages are in the list, thus you can use list::splice. Recently I submitted PR with write queue which does keeps queued messages in the list and puts them when data is sent, please take a look here #26932 

And please, no more locks. "send lock, write lock" - how this is supposed to be used?

**Comment by dillaman** [comment]
The problem is that ```Objecter``` can send the same message to multiple connections when the primary PG moves. If the message bufferlist has already been pulled out of the queue but is now waiting to send (or in-process of sending), we shouldn't complete the IO back through librados since the memory is still being referenced. 

An alternative to attempting to cancel messages would be to perhaps have an optional callback from ```::send_message``` when the message is completely written. Then ```Objecter``` can keep a count of in-flight memory references and delay the completion of the IO until all references are known to be released. The downside is that it would potentially drastically increase IO latency upon PG movement if the messenger's outbound writes aren't progressing (maybe not too much of an issue since it would most likely imply a non-graceful down OSD).

**Comment by rouming** [comment]
> we shouldn't complete the IO back through librados since the memory is still being referenced.

Under IO you mean Message object? If yes, I do not see any problems, since before submitting Message to the queue you just increase the reference on that Message preventing it from early free.

> An alternative to attempting to cancel messages would be to perhaps have an optional callback from ::send_message when the message is completely written.

I like callback approach very much. Message object has very short life-span and nobody reuses same message(or am I mistaken?). On hot IO write path atomic reference get/put affects performance, so plain callback gives possibility to reuse the same message and avoids atomic ref inc/dec. Once I played with the simple callback and that gives some perf gain: https://github.com/rouming/ceph/commit/3e34d9271ae104054044f2afcf8db655fc1f9fad

**Comment by dillaman** [comment]
> Under IO you mean Message object? If yes, I do not see any problems, since before submitting Message to the queue you just increase the reference on that Message preventing it from early free.

I mean ```Objecter``` shouldn't complete the ```Objecter::Op``` until after all references are completed (from a librbd point-of-view, these ops would be IOs from the C API). The ```Objecter``` will create branch new ```Message``` objects that are tied to the same ```Objecter::Op``` when it needs to redirect the message to a different OSD.

**Comment by rouming** [comment]
> I mean Objecter shouldn't complete the Objecter::Op until after all references are completed (

then of course you have to count messages, which reference same data, and then expect same number of completion callbacks. I vote for completion callbacks then! that is very well scalable approach which provides very good semantics without touching all the messenger internals, which frankly should not be aware about upper layers.

**Comment by dillaman** [comment]
@liyichao are you interested in working on the revised proposal to constrain the the changes to ```Objecter```? If not, I can take a stab at it this week.

**Comment by liyichao** [comment]
I am interested. I'm worried that if an OSD is down, will it block requests for a long time?

**Comment by dillaman** [comment]
> I am interested, just worried that if an OSD is down, will it block requests for a long time?

AFAIK, it would only block if the OSD w/ the destination PG goes down. In which case, IO would be staled regardless until the MONs update the OSD map. Once an OSD is flagged as down via the OSD map, the Objecter will move all ops to the homeless session, close the old session (and associated connection and blocked ops), and kick the homeless ops if they have a new home.  Worth a test, but it seems like it really shouldn't be any worse off.

**Comment by liyichao** [comment]
Then  the request will be stuck at the homeless ops ?

**Comment by dillaman** [comment]
> Then the request will be stuck at the homeless ops ?

Negative -- they should immediately be re-assigned to new sessions from the homeless session if the OSD map has an up (temporary) primary PG for the objects.

**Comment by liyichao** [comment]
OK, I will do this.

**Comment by liyichao** [comment]
@dillaman I have added the pr here https://github.com/ceph/ceph/pull/28106 , please have a look.

**Pull request closed without merging**